src/hotspot/share/services/memReporter.hpp

```diff
@@ -67,10 +67,19 @@ class MemReporterBase : public StackObj {
 
   // Convert diff amount in bytes to current reporting scale
   inline long diff_in_current_scale(size_t s1, size_t s2) const {
-    long amount = (long)(s1 - s2);
-    long scale = (long)_scale;
-    amount = (amount > 0) ? (amount + scale / 2) : (amount - scale / 2);
-    return amount / scale;
+    // Since size_t is unsigned, the amount >= 0 is always true.
+    // Therefore, shifting the amount half the scale down/up is
+    // decided based on the s1 and s2. Thus, s1 >= s2 means amount >= 0,
+    // and amount >= _scale/2 is equivalent to (amount - _scale / 2) >= 0.
+    size_t amount = s1 - s2;
+    if (s1 >= s2) {
+      assert((amount + _scale / 2) <= SIZE_MAX, "difference is gretaer than the upper limit.");
+      amount += _scale / 2;
+    } else {
+      assert(amount >= _scale / 2, "difference is less than the lower limit.");
+      amount -= _scale / 2;
+    }
+    return amount / _scale;
```

Comment:
I think the correct code should be like the following:
```
bool is_negative = false;
if (s1 < s2) {
  is_negative = true;
  swap(s1, s2);
}

size_t amount = s1 - s2;
assert(amount <= SIZE_MAX - _scale / 2, "size_t overflow");
amount = (amount + _scale / 2) / _scale;
if (is_negative) {
  assert(amount - 1  <= LONG_MAX, "cannot fit scaled diff into long);
  return (long)(-(long long)amount);
} else {
  assert(amount <= LONG_MAX, "cannot fit scaled diff into long);
  return (long)amount;
}
```

Reply:
New ways of handling wrap-around of size_t variables are implemented and are under tests.

Reply:
1) When amount is negative, it is a very large positive number as it is unsigned. 
2) amount +/- _scale/2 may wrap-around the boundaries.
3) In some cases, the actual negative amount should be returned. Expected by some tests.
 
My new code under test (All NMT tests passed locally, mach5 tier1-5 ongoing) is as follows:
```
inline long diff_in_current_scale(size_t s1, size_t s2) const {
    // _scale should not be 0, otherwise division by zero at return.
    assert(_scale != 0, "wrong scale");

    // Since size_t is unsigned, the amount >= 0 is always true.
    // Therefore, shifting the amount half the scale down/up is
    // decided based on the s1 and s2. Thus, s1 >= s2 means amount >= 0.
    // ((a + b) > a) where b != 0, is used to check wrap-around of adding two size_t variables.
    // ((a - b) < a) where b != 0, is used to check wrap-around of subtracting two size_t variables.
    size_t amount = s1 - s2;
    if (amount == 0){
      return 0L;
    }

    if (s1 >= s2) {
      if ((_scale / 2) != 0) {
        assert((amount + _scale / 2) > amount, "difference is greater than the upper limit.");
        amount += _scale / 2;
      }
      return (long)(amount / _scale);
    }

    if ((_scale / 2) != 0){
      assert((amount - _scale / 2) < amount, "difference is less than the lower limit.");
      amount -= _scale / 2;
    }
    return (long) amount / (long) _scale;
  }

```

Reply:
I don't get how this helps for deltas > 4gb if long is 4 and size_t is 8 bytes?